File: packages/mirascript/src/vm/lib/global/bit.ts
import type { VmAny } from '../../../index.js';
import { VmLib, expectNumber } from '../helpers.js';

/** 生成函数 */
function build(
    f: (x: number, y: number) => number,
    summary: string,
    examples: string[],
    y = '第二个操作数',
): VmLib<(str: VmAny, search: VmAny) => number> {
    return VmLib(
        (x, y) => {
            const a = expectNumber(0, x);
            const b = expectNumber(1, y);
            return f(a, b);
        },
        {
            summary,
            params: {
                x: { type: 'number', description: '第一个操作数' },
                y: { type: 'number', description: y },
            },
            returns: { type: 'number' },
            examples,
        },
    );
}

export const b_and = build((x, y) => x & y, '返回两个数的按位与', ['b_and(6, 3) // 2']);

export const b_or = build((x, y) => x | y, '返回两个数的按位或', ['b_or(5, 2) // 7']);

export const b_xor = build((x, y) => x ^ y, '返回两个数的按位异或', ['b_xor(5, 3) // 6']);

export const b_not = VmLib(
    (x) => {
        return ~expectNumber('x', x);
    },
    {
        summary: '返回一个数的按位取反',
        params: { x: { type: 'number', description: '操作数' } },
        returns: { type: 'number' },
        examples: ['b_not(0) // -1'],
    },
);

export const shl = build((x, y) => x << y, '返回第一个操作数左移指定的位数', ['shl(3, 2) // 12'], '位数');

export const sar = build((x, y) => x >> y, '返回第一个操作数右移指定的位数', ['sar(-8, 1) // -4'], '位数');

export const shr = build((x, y) => x >>> y, '返回第一个操作数无符号右移指定的位数', ['shr(8, 1) // 4'], '位数');

File: crates/python/mirascript/vm/lib/vm_global/bit.py
from mirascript.vm.lib._helpers import expect_number
import math

def to_int32(x):
    x = expect_number('x', x)
    if math.isnan(x):
        return 0
    
    x = int(x)
    return x 

def b_and(a, b):
    a = to_int32(a)
    b = to_int32(b)

    
    result =( a & b) 
    
    if result & 0x80000000:
        result -= 0x100000000
    
    
    return float(result)
  
def b_or(a, b):
    a = to_int32(a)
    b = to_int32(b)
    result = (a | b) & 0xFFFFFFFF
    if result & 0x80000000:
        result -= 0x100000000
    return float(result)

def b_not(a):
    a = to_int32(a)
    result = (~a) & 0xFFFFFFFF
    if result & 0x80000000:
        result -= 0x100000000
    
    return float(result)
  
def b_xor(a, b):
    a = to_int32(a)
    b = to_int32(b)
    result = (a ^ b) & 0xFFFFFFFF
    
    if result & 0x80000000:
        result -= 0x100000000
    
    
    return result
  
def shl(a, b):
    a = to_int32(a)
    b = to_int32(b)
    
    b &= 31  # 只保留低5位（0~31）
    a &= 0xFFFFFFFF  # 保持32位范围
    return ((a << b) | (a >> (32 - b))) & 0xFFFFFFFF

def sar(a, b):
    a = to_int32(a)
    b = to_int32(b)
    result = (a >> b) 
    
    
    
    return result
  
def shr(a, b):
    a = to_int32(a)
    b = to_int32(b)
    a = int(expect_number('a', a))
    b = int(expect_number('b', b))
    if a >= 0:
        return float(int(a) >> int(b))
    else:
        return float((int(a) + 0x100000000) >> int(b))